syntax/grammar.mjs

```diff
@@ -0,0 +1,495 @@
+import * as FTL from "./ast.mjs";
+import into from "./abstract.mjs";
+import {
+    after, always, and, between, char, charset, defer, either, eof, maybe, not,
+    regex, repeat, repeat1, sequence, string
+} from "../lib/combinators.mjs";
+import {
+    flatten, mutate, print, to_array, to_object, to_string
+} from "../lib/util.mjs";
+
+/* ------------------------------- */
+/* An FTL file defines a Resource. */
+export
+let Resource = defer(() =>
+    repeat(
+        either(
+            blank_line,
+            Entry.as("entry")))
+    .map(to_array)
+    .map(into(FTL.Resource)));
+
+export
+let Entry = defer(() =>
+    either(
+        Message,
+        Term,
+        either(
+            ResourceComment,
+            GroupComment,
+            Comment),
+        Junk));
+
+let Message = defer(() =>
+    sequence(
+        maybe(Comment).as("comment"),
+        Identifier.as("id"),
+        between(
+            maybe(inline_space),
+            char("=")),
+        either(
+            sequence(
+                Pattern.as("value"),
+                repeat(Attribute).as("attributes")),
+            sequence(
+                always(null).as("value"),
+                repeat1(Attribute).as("attributes"))),
+        line_end)
+    .map(flatten(1))
+    .map(to_object)
+    .map(into(FTL.Message)));
+
+let Term = defer(() =>
+    sequence(
+        maybe(Comment).as("comment"),
+        TermIdentifier.as("id"),
+        between(
+            maybe(inline_space),
+            char("=")),
+        Pattern.as("value"),
+        repeat(Attribute).as("attributes"),
+        line_end)
+    .map(to_object)
+    .map(into(FTL.Term)));
+
+let Attribute = defer(() =>
```

Comment:
How about moving that until after the pattern block? I find it's position here to be a bit of a break in reading the grammar and the ebnf.

Reply:
Good point. I'll move it into it's own comment group right above `Pattern`, because `Attribute` references it as well and I'd like to keep all `Expressions` right below `Pattern`.